icu4j/main/tests/core/src/com/ibm/icu/dev/test/message2/FromJsonTest.java

```diff
@@ -0,0 +1,412 @@
+                .build(),
+            new TestCase.Builder()
+                .pattern("{hello {(4.2) :number minimumFractionDigits=$foo}}")
+                .arguments(Args.of( "foo", 2f))
+                .expected("hello 4.20")
+                .build(),
+            new TestCase.Builder()
+                .pattern("{hello {(4.2) :number minimumFractionDigits=$foo}}")
+                .arguments(Args.of( "foo", "2"))
+                .expected("hello 4.20")
+                // errorsJs: ["invalid-type"]
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {(bar)} {bar {$foo}}")
+                .expected("bar bar")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {(bar)} {bar {$foo}}")
+                .arguments(Args.of( "foo", "foo"))
+                // expectedJs: "bar foo"
+                // It is undefined if we allow arguments to override local variables, or it is an error.
+                // And undefined who wins if that happens, the local variable of the argument.
+                .expected("bar bar")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar} {bar {$foo}}")
+                .arguments(Args.of( "bar", "foo"))
+                .expected("bar foo")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar :number} {bar {$foo}}")
+                .arguments(Args.of( "bar", 4.2))
+                .expected("bar 4.2")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar :number minimumFractionDigits=2} {bar {$foo}}")
+                .arguments(Args.of( "bar", 4.2))
+                .expected("bar 4.20")
+                .build(),
+            new TestCase.Builder().ignore("Maybe") // Because minimumFractionDigits=foo
+                .pattern("let $foo = {$bar :number minimumFractionDigits=foo} {bar {$foo}}")
+                .arguments(Args.of( "bar", 4.2))
+                .expected("bar 4.2")
+                .errors("invalid-type")
+                .build(),
+            new TestCase.Builder().ignore("Maybe. Function speciffic behavior.")
+                .pattern("let $foo = {$bar :number} {bar {$foo}}")
+                .arguments(Args.of( "bar", "foo"))
+                .expected("bar NaN")
+                .build(),
+            new TestCase.Builder()
+                .pattern("let $foo = {$bar} let $bar = {$baz} {bar {$foo}}")
+                .arguments(Args.of( "baz", "foo"))
+                // expectedJs: "bar foo"
+                // It is currently undefined if a local variable (like $foo)
+                // can reference a local variable that was not yet defined (like $bar).
+                // That is called hoisting and it is valid in JavaScript or Python.
+                // Not allowing that would prevent circular references.
```

Comment:
opinion: let's not allow that

Reply:
Mentioned in https://github.com/unicode-org/message-format-wg/issues/292
I proposed not to support hoisting, and I don't think there is much resistance to the idea.

Reply:
Added a link to the issue in the code.